cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,187 @@ def get_sbom_components(self) -> list[dict[str, str]]:
         packages = self._dependencies if self.lockfile_version == 1 else self._packages
         return [{"name": package.name, "version": package.version} for package in packages]
 
+    def get_dependencies_to_download(self) -> Dict[str, Dict[str, Optional[str]]]:
+        """Return a Dict of URL dependencies to download."""
+        packages = self._dependencies if self.lockfile_version == 1 else self._packages
+        return {
+            package.name: {
+                "version": package.version,
+                "url": package.resolved_url,
+                "integrity": package.integrity,
+            }
+            for package in packages
+            if "file:" not in package.resolved_url
+        }
+
+
+def _update_vcs_url_with_full_hostname(vcs: str) -> str:
+    """Update VCS URL with full hostname.
+
+    Transform github:kevva/is-positive#97edff6
+    into git+ssh://github.com/kevva/is-positive.git#97edff6
+
+    :param vcs: VCS URL to be modified with full hostname and file extension
+    :return: Updated VCS URL
+    """
+    host, _, path = vcs.partition(":")
+    namespace_repo, _, ref = path.rpartition("#")
+    suffix_domain = "org" if host == "bitbucket" else "com"
+
+    vcs = f"git+ssh://{host}.{suffix_domain}/{namespace_repo}.git"
+    if ref:
+        vcs = f"{vcs}#{ref}"
```

Comment:
I just realized that if the ref is missing, rpartition doesn't do what we expect
```python
In [1]: "no/ref".rpartition("#")
Out[1]: ('', '', 'no/ref')
```
(the `namespace_repo` would be empty, `ref` would be the namespace+repo)

This should probably use the regular from-the-left `partition`
or [`rsplit("#", 1)`](https://github.com/containerbuildsystem/cachito/blob/6b01c763018f7569d2aeabbd982e2381297fc7cb/cachito/workers/pkg_managers/general_js.py#L709) (and check if the result is `[path, ref]` or just `[path]`)

Reply:
Nice catch. I added simple unit test for missing ref.